### PR TITLE
type safe API

### PR DESCRIPTION
**Pull request by googley42**

### CODE REVIEW

File: dynamodb/src/main/scala/zio/dynamodb/Codec.scala [hunk 1]
```diff
@@ -372,6 +372,31 @@ private[dynamodb] object Codec {
 
   private[dynamodb] object Decoder extends GeneratedCaseClassDecoders {
 
+    sealed trait ContainerField
```
Comment (googley42) [inline dynamodb/src/main/scala/zio/dynamodb/Codec.scala:375]: I noticed a bug in decoding container fields when they are missing so (probably unwisely) decide to fix it in this PR

File: .gitignore [hunk 2]
```diff
@@ -478,3 +478,4 @@ shell.nix
 # direnv
 .envrc
 .direnv
+/.vscode/
```
Comment (googley42) [inline .gitignore:481]: after adding this ignore I will try removing the vscode files in a following PR

**Review by nhyne — APPROVED** [review]

File: dynamodb/src/it/scala/zio/dynamodb/LiveSpec.scala [hunk 3]
```diff
@@ -640,7 +660,21 @@ object LiveSpec extends DefaultRunnableSpec {
             } yield assert(updated)(equalTo(Some(Item(id -> second, number -> 2))))
           }
         },
-        testM("remove item from list") {
+        testM("remove an element from list") {
```
[the remark this answers is not in the document]
Comment (googley42) [reply dynamodb/src/it/scala/zio/dynamodb/LiveSpec.scala]: good spot, I changes the one below to test the `remove(index: Int)` case

File: dynamodb/src/main/scala/zio/dynamodb/ProjectionExpression.scala [hunk 4]
```diff
@@ -204,25 +85,506 @@ sealed trait ProjectionExpression { self =>
...
+
+    def ===[To2](that: ProjectionExpression[To2])(implicit refersTo: RefersTo[To, To2]): ConditionExpression = {
+      val _ = refersTo
+      ConditionExpression.Equals(
+        ProjectionExpressionOperand(self),
+        ConditionExpression.Operand.ProjectionExpressionOperand(that)
+      )
+    }
+
+    // TODO: think about using != instead
```
[the remark this answers is not in the document]
Comment (googley42) [reply dynamodb/src/main/scala/zio/dynamodb/ProjectionExpression.scala:374]: I had a quick look at this. It probably make sense just to replace `<>` with `!=` for compatibility with SQL but the changes are widespread so probably a candidate for a more focused PR

File: project/project/project/metals.sbt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+// DO NOT EDIT! This file is auto-generated.
```
[the remark this answers is not in the document]
Comment (googley42) [reply project/project/project/metals.sbt]: deleted both

**Pull request merged**